### PR TITLE
fix #1225 : allow the user to select an area in the OSM downloader

### LINKED ISSUE #1225: More interactive selection of area to download OSM data
# Problem

Current option to download OSM data from OSM downloader are to:
 - use the current map extent;
 - manually provide NSEW coordinates.

For some areas (see image), the extent might be too big (either horizontally and vertically).

screenshot from 2014-09-25 12 24 09

# Proposed solution

Allow the OSM downloader to have an interactive boundingbox selection option similar to GDAL_translate's Clipper UI in QGIS (see image).

screenshot from 2014-09-25 12 30 12

### PR DESCRIPTION
**Pull request by Gustry**
This allows the user to select an area by drawing a rectangle, in the OSM downloader.
Until you start drawing a rectangle, the extent is set to the map canvas when the window is showed.
You can click on a button to hide the window and start drawing a rectangle. If the rectangle is valid, the extent is changed.

### CODE REVIEW

**Comment by Gustry** [comment]
Even if jenkins failed, could someone review this PR ?

**Comment by akbargumbira** [comment]
test this please, jenkins

File: safe/gui/tools/osm_downloader_dialog.py [hunk 1]
```diff
@@ -174,15 +186,45 @@ def show_help(self):
         """Load the help text for the dialog."""
         show_context_help(self.help_context)
 
-    def update_extent(self):
-        """ Update extent value in GUI based from value in map."""
-        # Get the extent as [xmin, ymin, xmax, ymax]
-        extent = viewport_geo_array(self.iface.mapCanvas())
+    def update_extent(self, extent):
+        """ Update extent value in GUI based from an extent.
```
Comment (timlinux) [inline safe/gui/tools/osm_downloader_dialog.py]: """Update extent value in canvas using the supplied extent.

File: safe/gui/tools/osm_downloader_dialog.py [hunk 2]
```diff
@@ -174,15 +186,45 @@ def show_help(self):
...
+
+        :param extent: a list in the form [xmin, ymin, xmax, ymax] where all
```
Comment (timlinux) [inline safe/gui/tools/osm_downloader_dialog.py]: A list .... (docstring parameter descriptions should start with a capital letter and be fully formed sentances).

File: safe/gui/tools/osm_downloader_dialog.py [hunk 3]
```diff
@@ -174,15 +186,45 @@ def show_help(self):
...
+            coordinates provided are in Geographic / EPSG:4326.
+        :type: list
+        """
         self.min_longitude.setText(str(extent[0]))
         self.min_latitude.setText(str(extent[1]))
         self.max_longitude.setText(str(extent[2]))
         self.max_latitude.setText(str(extent[3]))
 
+    def update_extent_from_map_canvas(self):
+        """ Update extent value in GUI based from value in map.
```
Comment (timlinux) [inline safe/gui/tools/osm_downloader_dialog.py]: No space after """ (applies everywhere)

File: safe/gui/tools/osm_downloader_dialog.py [hunk 4]
```diff
@@ -419,3 +478,13 @@ def load_shapefile(self, feature_type):
             raise ImportDialogError(message)
 
         self.iface.addVectorLayer(path, feature_type, 'ogr')
+
+    def reject(self):
+        """Redefinition of the reject() method to remove the rectangle selection tool.
```
Comment (timlinux) [inline safe/gui/tools/osm_downloader_dialog.py]: Separate first sentence of docstring from detailed content with a newline.

File: safe/utilities/gis.py [hunk 5]
```diff
@@ -98,8 +98,11 @@ def extent_to_array(extent, source_crs, dest_crs=None):
     return geo_extent
 
 
-def viewport_geo_array(map_canvas):
-    """Obtain the map canvas current extent in EPSG:4326.
+def rectangle_geo_array(rectangle, map_canvas):
+    """Obtain the get_rectangle in EPSG:4326.
+
+    :param rectangle: A get_rectangle instance.
```
Comment (timlinux) [inline safe/utilities/gis.py]: remove get_

File: safe/utilities/gis.py [hunk 6]
```diff
@@ -98,8 +98,11 @@ def extent_to_array(extent, source_crs, dest_crs=None):
     return geo_extent
 
 
-def viewport_geo_array(map_canvas):
-    """Obtain the map canvas current extent in EPSG:4326.
+def rectangle_geo_array(rectangle, map_canvas):
+    """Obtain the get_rectangle in EPSG:4326.
```
Comment (timlinux) [inline safe/utilities/gis.py]: remove get_

**Comment by timlinux** [comment]
LGTM once my nitpicks are dealt with

File: safe/gui/tools/osm_downloader_dialog.py [hunk 7]
```diff
@@ -97,7 +99,17 @@ def __init__(self, parent=None, iface=None):
         if proxy is not None:
             self.network_manager.setProxy(proxy)
         self.restore_state()
-        self.update_extent()
+
+        # Setup the rectangle map tool
+        self.canvas = iface.mapCanvas()
+        self.rectangle_map_tool = RectangleMapTool(self.canvas)
+        self.rectangle_map_tool.rectangle_created.connect(self.update_extent_from_rectangle)
```
Comment (ismailsunni) [inline safe/gui/tools/osm_downloader_dialog.py]: Hi @Gustry 

This line 106 and 107 are violating PEP8. Perhaps, you want to run `make pep8` first to see if there is some PEP8 violations.

Thanks

Comment (Gustry) [inline safe/gui/tools/osm_downloader_dialog.py]: Oh, ok sorry. I will run that before the next time.

**Pull request merged**